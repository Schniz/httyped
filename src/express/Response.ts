export type Result<T> =
  | {
      success: false;
      errors: string[];
    }
  | { success: true; data: T };

export type Response<T> = {
  status: number;
  headers?: { [key: string]: string };
  body: Result<T>;
};

export function success<T>(data: T): Response<T> {
  return { status: 200, headers: {}, body: { success: true, data } };
}

export function response<T>(response: Response<T>): Response<T> {
  return response;
}
