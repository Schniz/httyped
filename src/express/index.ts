export { success, Response } from './Response';
export { TypedExpress } from './TypedExpress';
