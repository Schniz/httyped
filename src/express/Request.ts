import { RequestParamFetcher } from "./RequestParamFetcher";
import { RouteDefiner, RouteParams, Method } from "../RouteDefiner";
import * as t from "io-ts";
import * as E from "express";

export type Request<P extends string, RequestType> = {
  params: RouteParams<P>;
  body: RequestType;
};

export function make<
  I extends t.Any,
  O extends t.Any,
  P extends string,
  M extends Method
>(req: E.Request, rd: RouteDefiner<I, O, P, M>): Request<P, I> {
  return {
    params: new RequestParamFetcher(rd, req).getAll(),
    body: methodsWithoutBodies.has(rd.method) ? null : (req.body as t.TypeOf<I>)
  };
}

const methodsWithoutBodies: Set<Method> = new Set(["get"]);
