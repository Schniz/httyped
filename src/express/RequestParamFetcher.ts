import { RouteDefiner } from "../RouteDefiner";
import { Request } from "express";

export class RequestParamFetcher<ParamNames extends string> {
  routeDefiner: RouteDefiner<any, any, ParamNames, any>;
  request: Request;

  constructor(rd: RouteDefiner<any, any, ParamNames, any>, request: Request) {
    this.routeDefiner = rd;
    this.request = request;
  }

  get(name: ParamNames): string {
    return this.request.param(name);
  }

  getAll(): { [param in ParamNames]: string } {
    return this.request.params;
  }
}
