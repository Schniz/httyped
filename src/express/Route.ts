import * as t from "io-ts";
import { RouteDefiner } from "../RouteDefiner";
import { Response } from "./Response";
import { Request } from "./Request";
import bodyParser = require("body-parser");
import * as E from "express";
import { PathReporter } from "io-ts/lib/PathReporter";

export type RouteCallback<
  RequestBody,
  ResponseBody,
  ParamNames extends string
> = (
  params: Request<ParamNames, RequestBody>
) => Promise<Response<ResponseBody>>;

export class ServerRoute<
  RequestBody extends t.Any,
  ResponseBody extends t.Any,
  P extends string,
  M extends string
> {
  routeDefiner: RouteDefiner<RequestBody, ResponseBody, P, M>;
  callback: RouteCallback<RequestBody, ResponseBody, P>;

  constructor(opts: {
    routeDefiner: RouteDefiner<RequestBody, ResponseBody, P, M>;
    callback: RouteCallback<RequestBody, ResponseBody, P>;
  }) {
    this.routeDefiner = opts.routeDefiner;
    this.callback = opts.callback;
  }

  getMiddlewares() {
    const requestType = this.routeDefiner.requestType;
    if (!requestType) {
      return [];
    }
    const rt = requestType as t.Any;

    const parser = rt.name === "string" ? bodyParser.text() : bodyParser.json();

    return [parser, validate(rt)];
  }
}

const validate = (requestType: t.Any) => (
  req: E.Request,
  res: E.Response,
  next: E.NextFunction
) => {
  const body = req.body;
  const result = requestType.decode(body);
  const report = PathReporter.report(result);

  if (result.isRight()) {
    return next();
  } else {
    return res.status(406).json({
      success: false,
      errors: report
    });
  }
};
