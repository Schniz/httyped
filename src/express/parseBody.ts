import * as E from "express";
import * as t from "io-ts";
import * as bodyParser from "body-parser";
import { PathReporter } from "io-ts/lib/PathReporter";

export function parseBody<BodyType extends t.Any>(
  bodyType: BodyType
): ExpressMiddleware[] {
  const parser =
    bodyType.name === "string" ? bodyParser.text() : bodyParser.json();

  const validate = (req: E.Request, res: E.Response, next: E.NextFunction) => {
    const body = req.body;
    const result = bodyType.decode(body);
    const report = PathReporter.report(result);

    if (result.isRight()) {
      return next();
    } else {
      return res.status(406).json({
        success: false,
        errors: report
      });
    }
  };

  return [parser, validate];
}

type ExpressMiddleware = (
  req: E.Request,
  res: E.Response,
  next: E.NextFunction
) => void;
