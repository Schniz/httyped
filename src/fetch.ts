import * as t from "io-ts";
import { RouteDefiner, Meta } from "./RouteDefiner";
import nodeFetch from "node-fetch";
import { Response } from "./express/Response";

type Opts<M, Params, Body> = ([keyof Params] extends [never]
  ? { params?: Params }
  : {
      params: Params;
    }) &
  (M extends "get"
    ? {}
    : {
        body: [Body] extends [never] ? undefined : Body;
      });

export async function fetch<
  I extends t.Any,
  O extends t.Any,
  P extends string,
  M extends string
>(
  rd: RouteDefiner<I, O, P, M>,
  opts: Opts<M, Meta.Params<typeof rd>, Meta.RequestBodyType<typeof rd>>,
  rootURL: string
): Promise<Meta.ResponseType<typeof rd>> {
  const path = rd.toString(opts.params || ({} as Meta.Params<typeof rd>));
  const response = await nodeFetch(`${rootURL}${path}`, {
    method: rd.method,
    ...(rd.method !== "get" && {
      headers: {
        "Content-Type":
          rd.requestType.name === "string" ? "text/plain" : "application/json"
      },
      body:
        rd.requestType.name === "string" ? opts.body : JSON.stringify(opts.body)
    })
  });

  return response.json();
}

export function fetcher<
  I extends t.Any,
  O extends t.Any,
  P extends string,
  M extends string
>(
  rd: RouteDefiner<I, O, P, M>,
  rootURL: string = "/"
): (
  opts: Opts<M, Meta.Params<typeof rd>, Meta.RequestBodyType<typeof rd>>
) => Promise<Response<Meta.ResponseType<typeof rd>>["body"]> {
  return opts => fetch(rd, opts, rootURL);
}
