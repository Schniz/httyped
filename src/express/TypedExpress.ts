import * as E from "express";
import * as t from "io-ts";
import { RouteDefiner, Method } from "../RouteDefiner";
import { make as makeRequest } from "./Request";
import { RouteCallback, ServerRoute } from "./Route";

export class TypedExpress {
  private readonly app: E.Router;
  private routes: ServerRoute<any, any, any, any>[] = [];

  static of(app: E.Router) {
    return new TypedExpress(app);
  }

  constructor(app: E.Router) {
    this.app = app;
  }

  route<
    Params extends string,
    RouteResult extends t.Any,
    RequestType extends t.Any
  >(
    routeDefiner: RouteDefiner<RequestType, RouteResult, Params, Method>,
    callback: RouteCallback<
      t.TypeOf<RequestType>,
      t.TypeOf<RouteResult>,
      Params
    >
  ) {
    const route = new ServerRoute({
      callback,
      routeDefiner
    });
    this.routes.push(route);

    const routeString = route.routeDefiner.toRoutingString();
    this.app[route.routeDefiner.method as "get" | "post"](
      routeString,
      route.getMiddlewares(),
      async (req: E.Request, res: E.Response, next: E.NextFunction) => {
        try {
          const typedRequest = makeRequest(req, route.routeDefiner);
          const { status, body, headers } = await route.callback(typedRequest);
          for (const [key, value] of Object.entries(headers || {})) {
            res.header(key, value);
          }
          res.status(status).json(body);
        } catch (e) {
          next(e);
        }
      }
    );
  }

  listRoutes() {
    const tableData = this.routes.map(route => {
      return route.routeDefiner.toJSON();
    });

    console.log("Routes:");
    console.table(tableData);
  }
}
