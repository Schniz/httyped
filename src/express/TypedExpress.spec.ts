import * as t from "io-ts";
import fetch from "node-fetch";
import express, { Express } from "express";
import { TypedExpress } from "./TypedExpress";
import { success } from "./Response";
import { RouteDefiner } from "../RouteDefiner";
import { Server } from "http";
import { AddressInfo } from "net";

let server: Server | null = null;

afterEach(() => {
  if (server) {
    server.close();
    server = null;
  }
});

test.only("Simple application works", async () => {
  const application = express();
  const app = new TypedExpress(application);

  app.and(RouteDefiner.get.returns(t.string).fixed("hello"), async () =>
    success("Hello world!")
  );

  const host = getHost(application);

  const result = await fetch(`${host}/hello`).then(x => x.json());

  expect(result).toEqual({ success: true, data: "Hello world!" });
});

test("Parameters work", async () => {
  const application = express();
  const app = new TypedExpress(application);

  const Game = t.type({ name1: t.string, name2: t.string });
  app.and(
    RouteDefiner.get
      .returns(Game)
      .param("name1")
      .fixed("vs")
      .param("name2"),
    async req => success({ name1: req.params.name1, name2: req.params.name2 })
  );

  const host = getHost(application);

  const result = await fetch(`${host}/gal/vs/the world`).then(x => x.json());

  expect(Game.decode(result.data).isRight()).toBe(true);
  expect(result.data).toEqual({ name1: "gal", name2: "the world" });
});

function getHost(app: Express) {
  server = app.listen(0);
  const port = (server.address() as AddressInfo).port;
  return `http://localhost:${port}`;
}
