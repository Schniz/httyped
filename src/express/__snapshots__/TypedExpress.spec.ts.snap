// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Types are inferred correctly 1`] = `
Object {
  "RouteDefiner.get.param(\\"name\\")": "(req: Request<\\"name\\", unknown>) => Promise<Response<string>>",
  "RouteDefiner.post.reads(t.string).param(\\"name\\")": "(req: Request<\\"name\\", string>) => Promise<Response<string>>",
  "RouteDefiner.post.reads(t.string).returns(t.type({ name: t.string })).param(\\"name\\")": "(req: Request<\\"name\\", string>) => Promise<Response<{ name: string; }>>",
}
`;
